src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -117,12 +117,16 @@ public HttpConnectionPool(HttpConnectionPoolManager poolManager, HttpConnectionK
 
             if (_host != null)
             {
+                int posColon = -1;
+                bool isHostTypeIPv6 = (posColon = _host.IndexOf(':')) != -1 && _host.IndexOf(':', posColon+1) != -1;
+                string hostAddress = isHostTypeIPv6 ? "[" + _host + "]" : _host;
+
```

Comment:
Can we just use Uri.CheckHostName here?

Reply:
Alternatively, can we just go back to checking for HostNameType == IPv6 in GetConnectionKey, but use uri.Host instead of uri.IdnHost in that case?  I think that will return the address with [], right?

I'm kinda queasy about doing our own parsing here.  Uri does all sorts of stuff for us already, let's just use it if possible instead of trying to duplicate the logic ourselves.

Reply:
I will also point out that retrieving parts of the Uri (e.g. hostname) is already a bit of a perf bottleneck.  Uri.IdnHost is allocating a string.  I believe we also end up allocating later when we retrieve path/query etc.  

We should look at optimizing this better, but not for 2.1.

In the meantime let's just do something simple and correct, and we can optimize interaction with Uri in the future.

Reply:
> can we just go back to checking for HostNameType == IPv6 in GetConnectionKey, but use uri.Host instead of uri.IdnHost in that case?

WinHttp [clearly documented](https://github.com/dotnet/corefx/issues/28557#issuecomment-377085162) that it will only accept Punycode host name. Changing to uri.Host may bring regression for SocketsHttpHandler, and I think that's risky for 2.1.

> I think that will return the address with [], right?

Yes, uri.Host will return the address with [].

> I'm kinda queasy about doing our own parsing here.

I will change to `Uri.CheckHostName`.

Reply:
> WinHttp clearly documented that it will only accept Punycode host name. Changing to uri.Host may bring regression for SocketsHttpHandler, and I think that's risky for 2.1.

I'm not suggesting we change it in *all* cases.  Obviously that would break Punycode handling.  Just in the case where it's an IP address.

Reply:
I think that should work. I will push a new commit.